wandb/sdk/lib/filenames.py

```diff
@@ -29,15 +29,30 @@ def is_wandb_file(name: str) -> bool:
 
 
 def filtered_dir(
-    root: str, include_fn: Callable[[str], bool], exclude_fn: Callable[[str], bool]
+    root: str,
+    include_fn: Union[Callable[[str, str], bool], Callable[[str], bool]],
+    exclude_fn: Union[Callable[[str, str], bool], Callable[[str], bool]],
 ) -> Generator[str, None, None]:
     """Simple generator to walk a directory."""
+    import inspect
+
+    # compatibility with old API, which didn't pass root
```

Comment:
Unfortunately required since this is public API.  Any existing callers would break if we call blindly with two parameters.

Reply:
why do you need to add the root in the include/exclude functions?

Reply:
AFAICT the original intent of this logic is to prevent logging the files under the `wandb` metadata directory, but it was never implemented in a way that could detect that correctly.  We need the root because we ONLY want to exclude `<root>/.wandb/*` and `<root>/wandb/*` and nothing else.  The new unit test probes the new behavior exhaustively.

It's not safe to change the semantics of the first parameter to be a relative path from root because existing uses might break if they were previously checking for something in the prefix, e.g. `exclude_fn = lambda path: path.startswith('/home/project')` where root might be `/home/project/train` -- quite contrived but possible because this was public API.  Providing a second argument to the callback is a generally safe change as long as we inspect the provided callback's signature and dispatch appropriately.  This won't break existing uses, and enables us to call `os.path.relpath(path, root)` inside the default `exclude_fn`.

Reply:
Do we want eventually to have the api have both?

Reply:
hmm, I'm not sure I follow.  With this change, the API would support both forms

Reply:
sorry i meant would we want to have just the new api i.e. path and root as arguments or keep both options... since you could define any functions as the second one, by just passing the root provided. But it's really not an issue, just tried to understand the api.

Reply:
oh, I see. I was going for parallel change here, but did not see any way to annotate the 1-parameter form as deprecated.  Eventually I think we should hide that from the public API and just keep the hidden compatibility shim